tests/test_circuit.py

```diff
@@ -202,3 +214,59 @@ def test_merge_different_batchsize(batch_size: int) -> None:
 def test_ham_expect_fail() -> None:
     ops = [pyq.Z(0), pyq.RY(1, "theta_1")]
     ham = pyq.Hamiltonian(ops)
+
+
+@pytest.mark.parametrize("n_qubits", torch.randint(low=2, high=5, size=(5, 1)))
+def test_noise_circ(n_qubits: int) -> None:
+    GATES: list = [X, Y, Z, I, H]
+    NOISE_GATES: list = [
+        BitFlip,
+        PhaseFlip,
+        PauliChannel,
+        Depolarizing,
+        AmplitudeDamping,
+        PhaseDamping,
+        GeneralizedAmplitudeDamping,
+    ]
+    batch_size: int = torch.randint(low=1, high=5, size=(1,)).item()
+    input_state: Tensor = random_state(n_qubits, batch_size)
+    op1, op2 = random.choice(GATES), random.choice(GATES)
+    op1 = op1(torch.randint(0, n_qubits, (1,)).item())
+    op2 = op2(torch.randint(0, n_qubits, (1,)).item())
+    noise_1, noise_2 = random.choice(NOISE_GATES), random.choice(NOISE_GATES)
+    if noise_1 == PauliChannel:
```

Comment:
This test is good but it deserves to be reorganised I think. All this prep work until l. 262 should go in fixtures IMHO. You can have fixture that return a tuple of parameters needed next, _i.e_ `(n_qubits, op1, op2, noise_1, noise_2)` (possibly few more that I am missing here or they can also be other fixtures). Then all your test will do is circuit construction and a bunch of assertions which make it more readable and comprehensible.

Reply:
very neat

Reply:
Modified.